File: languages/java/src/main/java/com/bloock/sdk/entity/identity/CredentialBuilder.java
package com.bloock.sdk.entity.identity;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.*;
import com.bloock.sdk.entity.key.Key;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Helps construct credentials by specifying various attributes.
 */
public class CredentialBuilder {
  private final String schemaId;
  private final String holderKey;
  private final Long expiration;
  private final int version;
  private KeysEntities.Key key;
  private final Config.ConfigData configData;

  private final List<IdentityEntities.BooleanAttribute> booleanAttributes;
  private final List<IdentityEntities.DateAttribute> dateAttributes;
  private final List<IdentityEntities.DateTimeAttribute> datetimeAttributes;
  private final List<IdentityEntities.StringAttribute> stringAttributes;
  private final List<IdentityEntities.IntegerAttribute> integerAttributes;
  private final List<IdentityEntities.DecimalAttribute> decimalAttributes;

  /**
   * Creates a new CredentialBuilder instance with the specified parameters.
   * @param issuer
   * @param schemaId
   * @param holderKey
   * @param expiration
   * @param version
   * @param configData
   */
  public CredentialBuilder(
      Issuer issuer,
      String schemaId,
      String holderKey,
      Long expiration,
      int version,
      Config.ConfigData configData) {
    this.schemaId = schemaId;
    this.holderKey = holderKey;
    this.expiration = expiration;
    this.version = version;
    this.key = issuer.getKey().toProto();
    this.configData = configData;

    this.booleanAttributes = new ArrayList<>();
    this.dateAttributes = new ArrayList<>();
    this.datetimeAttributes = new ArrayList<>();
    this.stringAttributes = new ArrayList<>();
    this.integerAttributes = new ArrayList<>();
    this.decimalAttributes = new ArrayList<>();
  }

  /**
   * Adds a string attribute to the CredentialBuilder.
   * @param key
   * @param value
   * @return
   */
  public CredentialBuilder withStringAttribute(String key, String value) {
    this.stringAttributes.add(new StringAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds an integer attribute to the CredentialBuilder.
   * @param key
   * @param value
   * @return
   */
  public CredentialBuilder withIntegerAttribute(String key, Long value) {
    this.integerAttributes.add(new IntegerAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds a decimal attribute to the CredentialBuilder.
   * @param key
   * @param value
   * @return
   */
  public CredentialBuilder withDecimalAttribute(String key, double value) {
    this.decimalAttributes.add(new DecimalAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds a boolean attribute to the CredentialBuilder.
   * @param key
   * @param value
   * @return
   */
  public CredentialBuilder withBooleanAttribute(String key, boolean value) {
    this.booleanAttributes.add(new BooleanAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds a date attribute to the CredentialBuilder.
   * @param key
   * @param value
   * @return
   */
  public CredentialBuilder withDateAttribute(String key, LocalDate value) {
    this.dateAttributes.add(new DateAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds a datetime attribute to the CredentialBuilder.
   * @param key
   * @param value
   * @return
   */
  public CredentialBuilder withDatetimeAttribute(String key, LocalDateTime value) {
    this.datetimeAttributes.add(new DatetimeAttribute(key, value).toProto());
    return this;
  }

  /**
   * Creates and returns a Credential using the specified attributes.
   * @return
   * @throws Exception
   */
  public CredentialReceipt build() throws Exception {
    Bridge bridge = new Bridge();

    Identity.CreateCredentialRequest req =
        Identity.CreateCredentialRequest.newBuilder()
            .setConfigData(this.configData)
            .setSchemaId(this.schemaId)
            .setHolderDid(this.holderKey)
            .setExpiration(this.expiration)
            .setVersion(this.version)
            .setKey(this.key)
            .addAllStringAttributes(this.stringAttributes)
            .addAllIntegerAttributes(this.integerAttributes)
            .addAllDecimalAttributes(this.decimalAttributes)
            .addAllBooleanAttributes(this.booleanAttributes)
            .addAllDateAttributes(this.dateAttributes)
            .addAllDatetimeAttributes(this.datetimeAttributes)
            .build();

    Identity.CreateCredentialResponse response = bridge.getIdentity().createCredential(req);

    if (response.getError() != Shared.Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return CredentialReceipt.fromProto(response.getCredentialReceipt());
  }
}

File: languages/js/src/entity/identity_v2/credential_builder.ts
import { BloockBridge } from "../../bridge/bridge";
import { ConfigData } from "../../bridge/proto/config";
import { CreateCredentialRequestV2 } from "../../bridge/proto/identity_v2";
import { BooleanAttribute } from "./boolean_attribute";
import { CredentialReceipt } from "./credential_receipt";
import { DateAttribute } from "./date_attribute";
import { DateTimeAttribute } from "./datetime_attribute";
import {
  BooleanAttributeV2,
  DateAttributeV2,
  DateTimeAttributeV2,
  DecimalAttributeV2,
  IntegerAttributeV2,
  StringAttributeV2
} from "../../bridge/proto/identity_entities_v2";
import { Signer as SignerProto } from "../../bridge/proto/authenticity_entities";
import { Signer } from "../authenticity/signer";
import { DecimalAttribute } from "./decimal_attribute";
import { IntegerAttribute } from "./integer_attribute";
import { StringAttribute } from "./string_attribute";

/**
 * Helps construct credentials by specifying various attributes.
 */
export class CredentialBuilder {
  schemaId: string;
  issuerDid: string;
  holderDid: string;
  expiration: number;
  version: number;
  signer?: SignerProto;
  configData: ConfigData;

  stringAttributes: StringAttributeV2[];
  integerAttributes: IntegerAttributeV2[];
  decimalAttributes: DecimalAttributeV2[];
  booleanAttributes: BooleanAttributeV2[];
  dateAttributes: DateAttributeV2[];
  dateTimeAttributes: DateTimeAttributeV2[];

  /**
   * Creates a new CredentialBuilder instance with the specified parameters.
   * @param schemaId 
   * @param issuerDid 
   * @param holderDid 
   * @param expiration 
   * @param version 
   * @param configData 
   */
  constructor(
    schemaId: string,
    issuerDid: string,
    holderDid: string,
    expiration: number,
    version: number,
    configData: ConfigData
  ) {
    this.schemaId = schemaId;
    this.issuerDid = issuerDid;
    this.holderDid = holderDid;
    this.expiration = expiration;
    this.version = version;
    this.configData = configData;

    this.signer = undefined;
    this.stringAttributes = [];
    this.integerAttributes = [];
    this.decimalAttributes = [];
    this.booleanAttributes = [];
    this.dateAttributes = [];
    this.dateTimeAttributes = [];
  }

  /**
   * Adds a string attribute to the CredentialBuilder.
   * @param key 
   * @param value 
   * @returns 
   */
  public withStringAttribute(key: string, value: string): CredentialBuilder {
    this.stringAttributes.push(new StringAttribute(key, value));
    return this;
  }

  /**
   * Adds an integer attribute to the CredentialBuilder.
   * @param key 
   * @param value 
   * @returns 
   */
  public withIntegerAttribute(key: string, value: number): CredentialBuilder {
    this.integerAttributes.push(new IntegerAttribute(key, value));
    return this;
  }

  /**
   * Adds a decimal attribute to the CredentialBuilder.
   * @param key 
   * @param value 
   * @returns 
   */
  public withDecimalAttribute(key: string, value: number): CredentialBuilder {
    this.decimalAttributes.push(new DecimalAttribute(key, value));
    return this;
  }

  /**
   * Adds a boolean attribute to the CredentialBuilder.
   * @param key 
   * @param value 
   * @returns 
   */
  public withBoleanAttribute(key: string, value: boolean): CredentialBuilder {
    this.booleanAttributes.push(new BooleanAttribute(key, value));
    return this;
  }

  /**
   * Adds a date attribute to the CredentialBuilder.
   * @param key 
   * @param value 
   * @returns 
   */
  public withDateAttribute(key: string, value: Date): CredentialBuilder {
    this.dateAttributes.push(new DateAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds a datetime attribute to the CredentialBuilder.
   * @param key 
   * @param value 
   * @returns 
   */
  public withDateTimeAttribute(key: string, value: Date): CredentialBuilder {
    this.dateTimeAttributes.push(new DateTimeAttribute(key, value).toProto());
    return this;
  }

  /**
   * Sets the signer for the CredentialBuilder.
   * @param signer 
   * @returns 
   */
  public withSigner(signer: Signer): CredentialBuilder {
    this.signer = signer.toProto();
    return this;
  }

  /**
   * Creates and returns a Credential using the specified attributes.
   * @returns 
   */
  async build(): Promise<CredentialReceipt> {
    const bridge = new BloockBridge();

    const req = CreateCredentialRequestV2.fromPartial({
      configData: this.configData,
      schemaId: this.schemaId,
      issuerDid: this.issuerDid,
      holderDid: this.holderDid,
      expiration: this.expiration,
      version: this.version,
      signer: this.signer,
      stringAttributes: this.stringAttributes,
      integerAttributes: this.integerAttributes,
      decimalAttributes: this.decimalAttributes,
      booleanAttributes: this.booleanAttributes,
      dateAttributes: this.dateAttributes,
      datetimeAttributes: this.dateTimeAttributes
    });

    return bridge
      .getIdentityV2()
      .CreateCredential(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return CredentialReceipt.fromProto(res.credentialReceipt!);
      });
  }
}
